### PR TITLE
Remove duplicate fields

### LINKED ISSUE #358: gents field generation creates duplicate members
if an untyped member is assigned multiple times, it creates multiple properties with the same name and different initialized values.

input:
```
/**
 * @constructor
 */
var A = function() {
  this.n = 5;
};

A.prototype.foo = function() {
  this.n = 12;
};
```

output:
```
class A {
  n: any = 5;
  n: any = 12;
  foo() {}
}
```

expected output:
```
class A {
  n: number = 5;

  foo() {
    this.n = 12;
  }
}
```

### PR DESCRIPTION
**Pull request by gmoothart**
This fixes a bug that would cause gents to generate duplicate fields and promote some assignments to field initializers inappropriately.

It changes behavior to only look in the constructor for fields. With some care to remove duplicates and leave the assignments in-place it would be possible to promote fields in other methods too.

fixes #358

### CODE REVIEW

**Comment by gmoothart** [comment]
This branch is based off https://github.com/angular/clutz/pull/364; only review the second commit for this PR (is there a better way to do this in github?

File: src/test/java/com/google/javascript/gents/singleTests/fields.js [hunk 1]
```diff
@@ -11,6 +11,11 @@ var A = function(a) {
   let y = 1;
   /** @type {number} */
   this.z = y + 1;
+
+  this.w;
+  this.w.bar = 'bar';
+
+  this.n = 13;
```
Comment (mprobst) [inline src/test/java/com/google/javascript/gents/singleTests/fields.js]: can you add a test where the constructor mentions `this.n` twice, e.g. assigns `12` into it on the next line?

Comment (gmoothart) [inline src/test/java/com/google/javascript/gents/singleTests/fields.js]: :cry: You are right, we need to de-duplicate anyway. I'd like to do that in a separate PR, though.

Comment (gmoothart) [inline src/test/java/com/google/javascript/gents/singleTests/fields.js]: Actually, never mind. To do this right I need a different approach so I'm going to just do that directly.

**Comment by gmoothart** [comment]
@mprobst PTAL; new approach.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 2]
```diff
@@ -133,44 +136,61 @@ public void visit(NodeTraversal t, Node n, Node parent) {
    * Converts fields declared internally inside a class using the "this" keyword.
    */
   private class FieldOnThisConverter extends AbstractPostOrderCallback {
+    // Map from class node to its field names
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: `/** JSDoc */`

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 3]
```diff
@@ -133,44 +136,61 @@ public void visit(NodeTraversal t, Node n, Node parent) {
...
-          Node fnNode = NodeUtil.getEnclosingFunction(n);
-          String fnName = fnNode.getParent().getString();
+        Node fnNode = NodeUtil.getEnclosingFunction(n);
+        String fnName = fnNode.getParent().getString();
+
+        // TODO(gmoothart): in many cases we should be able to infer the type from the rhs if there
+        // is no jsDoc
+
+        // Handle parameter properties when we are in the constructor and have a
+        // declaration of the form this.name = name;
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: add how you handle them? I.e. "by converting them to private/public constructor parameters" or something like that?
